README.md
# homebridge-eDomoticz
A fully-fledged up-to-date Homebridge-Plugin
for use with [Homebridge](https://github.com/nfarina/homebridge) v0.2.1+
and [Domoticz](https://github.com/domoticz/domoticz)

##Supports:
###Standard HomeKit Types:
- Sockets (on/off) - Domoticz SwitchTypeVal: 0
- Lamps (on/off) - Domoticz SwitchTypeVal: 0
- Contact Sensors - Domoticz SwitchTypeVal: 2
- Blinds - Domoticz SwitchTypeVal: 3
- Smoke Detectors - Domoticz SwitchTypeVal: 5
- Blinds (inverted) - Domoticz SwitchTypeVal: 6
- Lamps (dimmer) - Domoticz SwitchTypeVal: 7
- Motion Sensors - Domoticz SwitchTypeVal: 8
- Push Switches -  Domoticz SwitchTypeVal: 9
- Lock Mechanisms - Domoticz SwitchTypeVal: 11
- Blinds (%) - Domoticz SwitchTypeVal: 13
- Blinds (& inverted) - Domoticz SwitchTypeVal: 16

##Provides:
###Custom HomeKit Types:
- General kWh power meters - Types: General, Current; SubType: kWh, mapped to Eve chars where possible
- CurrentCost USB power meter - Type: Usage, SubType: Electric, mapped to Eve chars where possible
- P1 Smart Meter (Electric & Gas), mapped to Eve chars where possible
- EvoHome** / OpenTherm Thermostat support - Types: Heating, Thermostat; SubTypes: Zone, SetPoint
- YouLess Meter (Current, Total and Today Total Consumption) - Type: YouLess Meter; SubType: YouLess counter, mapped to Eve chars where possible
- General Usage % meters (eg: Motherboard Sensors Hardware Device - CPU %, Mem %, HDD % etc) - Type: General; SubType: Percentage
- Temperature, Temp + Humidity, Temp + Humidity + Baro (Current Temperature, Current Humidity, Current Pressure in hPA) - Type: Temp, Temp + Humidty, Temp + Humidity + Baro [id'd as Eve Weather]
- DarkSkies Virtual Weather Station Sensors (Wind, Solar Radiation, Rainfall, Visibility, Barometer [id'd as Eve Weather])

** assumes the EvoHome has been setup according to [this script method](https://www.domoticz.com/wiki/Evohome#Scripting_for_RFG100).

###Todo:
- [ ] homebridge [plugin 2.0](https://github.com/nfarina/homebridge/pull/497) support
- [x] MQTT-based realtime updates
- [x] Hue/RGB
- [x] Blinds
- [x] m3 (gas usage)
- [x] Motion sensors
- [x] Smoke Detectors
- [x] Brightness/Dimming
- [x] kWh (electricity usage)
- [x] General % usage
- [x] Humidity
- [x] Pressure
- [x] YouLess Meter
- [x] Open/Closed contact sensors
- [ ] ...more sensors (ongoing)!

## Installation
```
sudo npm install -g homebridge-edomoticz
```

## Update
```
sudo npm update -g homebridge-edomoticz
```

## Configuration

~/.homebridge/config.json example:
```
{
    "bridge": {
        "name": "Homebridge",
        "username": "CC:21:3E:E4:DE:33",
        "port": 51826,
        "pin": "031-45-154"
    },
    "description": "Configuration file for (e)xtended Domoticz platform.",
    "platforms": [
        {
            "platform": "eDomoticz",
            "name": "eDomoticz",
            "server": "127.0.0.1",
            "port": "8080",
            "ssl": 0,
            "roomid": 0,
            "mqttenable": 1,
            "mqttserver": "127.0.0.1",
            "mqttport": "1883",
            "mqttauth": 0,
            "mqttuser": "",
            "mqttpass": ""
        }
    ],
    "accessories": []
}
```

## Tips

### Authentication
If Domoticz is set up to use basic or form login authentication, set "server":"user:pass@ip" within config.json. The plugin will internally extract the username and password, base64 encode it and send it as a http authorization header whenever it talks to your Domoticz server.

### SSL
Set "ssl":1 in config.json to turn on SSL (ie: server connects with https:// rather than http://). You will need to specify your SSL port - usually "port":"443" by default.

### Issues pairing to Homebridge when you have a lot of Domoticz sensors...
If you have more than 100 devices in Domoticz, you need to limit the number of devices exposed to HomeKit (HomeKit only supports 100 Accessories on a single bridge - whilst we could combine multiple sensors into a single homekit accessory within the plugin, the possible combinations out there are endless, so we won't).

Therefore, to reduce the number of devices exposed from Domoticz, create a roomplan within Domoticz via Setup > More Options > Plans > roomplan. Add only the devices you wish to be exposed to HomeKit to this new roomplan within Domoticz, and then get it's roomidx number. Set "roomid" in your config.json file to this room number.

### Is my <<some accessory>> supported??
See [Domoticz API Reference](https://www.domoticz.com/wiki/Domoticz_API/JSON_URL's#Retrieve_status_of_specific_device) - query your device as per the instructions there, and if your device’s SwitchTypeVal isn't in the 'Supports:' list or Type/SubType aren’t in the ’Provides:' list above then it'll just appear as an On/Off switch. [Open a new issue](https://github.com/PatchworkBoy/homebridge-eDomoticz/issues/new) including the output from the json api and I’ll get look into supporting that particular device more fully!

### What does the Override slider represent on the EvoHome Thermostat?
Override-Until time in minutes from the current time. Allows setting an override-until time upto 8 hours in the future. Setting this slider to 0 will set the heating mode to Auto. Setting it to 481 will set the override as a PermanentOverride.

### Logging
Complies with Homebridge's native logging & debugging methodology - see https://github.com/nfarina/homebridge/wiki/Basic-Trouble-Shooting


CHANGELOG
         ____                        _   _
     ___|  _ \  ___  _ __ ___   ___ | |_(_) ________
    / _ | | | |/ _ \| '_ ` _ \ / _ \| __| |/ __|_  /
   |  __| |_| | (_) | | | | | | (_) | |_| | (__ / /
    \___|____/ \___/|_| |_| |_|\___/ \__|_|\___/___|
       www.npmjs.com/package/homebridge-edomoticz

   A Platform Plugin for HomeBridge by Marci & TheRamon
           [http://twitter.com/marcisshadow] 
           [http://domoticz.com/forum/memberlist.php?mode=viewprofile&u=10884]
 
V1.0.1 - see https://github.com/PatchworkBoy/homebridge-edomoticz/commits/master
V1.0.0 - MQTT Merge
V0.2.3 - 2016/10/26
    - MQTT Integration
    - Switch and dimmer values caching
    - Support for Domoticz Security panel
    - Dedicated waterflow handling
    - Improved response times
    - Improved handling of characteristics (iOS 10)
 V0.2.2 - 2016/08/19
    - Release Candidate for MQTT Integration
 V0.2.1 - 2016/06/21
    - Added handling for sensors of type 'Lux' - cheers @100hz
    - Hue/RGB support - cheers @xTheRamon
    - Improved BlindSupport (iOS10 compatible) - cheers @xTheRamon
 V0.1.32 - 2016/04/04
    - Add Occupancy Sensor support - power switches in Domotics
      featuring the word "Occupied" (case sensitive) in the Name
      will be created in HomeKit as an Occupancy Sensor
 V0.1.31 - 2016/04/01
    - Revert blind support
 V0.1.28 - 2016/04/01
    - Tweak to EvoHome support
    - Add Blind (switchType 3) support
 V0.1.27 - 2016/03/12
    - Tweak to EvoHome support
 V0.1.26 - 2016/03/02
    - Added CurrentCost USB support (introduced to stable in v3.4834)
    - Type: Usage, SubType: Electric
 V0.1.25 - 2016/02/26
    - Revision history moved to CHANGELOG
 V0.1.24 - 2016/02/26
    - fixed batterysensor for Motion Detectors (was: get before it was added = err thrown)
    - merge @gerard33/patch-1 pull request: Add smoke detector status Normal
 V0.1.23 - 2016/02/20
    - merge in bugfixes by @EddyK69 - commit/1ae1020146b5761f1aaa9bd69b6675210a777290
    - consistent that/this
    - removed some fluff
    - globally switched roundToHalf to OneDP
 V0.1.21 & 22 - 2016/02/20
    - More work on Thermostat / SetPoint support
 V0.1.20 - 2016/02/19
    - Dimmers reflect power state
 V0.1.19 - 2016/02/19
    - Corrected UV Type checking (@EddyK69)
 V0.1.16 - 18 - 2016/02/19
    - added P1 Smart Meter Energy subtype support
    - more work on Evohome
 V0.1.14 & 15 - 2016/02/19
    - added P1 Smart Meter kWh & Gas type / subtype support
    - added UVN800 support (@EddyK69)
    - added preliminary EvoHom / OpenTherm Gateway Thermostat / SetPoint support
 V0.1.11 & 12 - 2016/02/10
    - rewritten switch detection to use SwitchTypeVal=0>17
    - de-verbosed object selectors & jshint'd end-to-end
    - added: smoke detector, motion detector
    - fixed: dimmmer allocation, contact sensors
 V0.1.10 - 2016/02/08
    - more work relating to dimmers and contact sensors
 V0.1.9 - 2016/02/07
    - hopefully fixes an oopsie related to brightness on EVERY switch.
 V0.1.8 - 2016/02/07
    - add support for SwitchType: Contact (basic open / closed)
 V0.1.6 - 2016/02/07
    - add another identifier for Current
 V0.1.5 - 2016/02/07
    - add another identifier for a dimming light
 V0.1.4 - 2016/02/06
    - removed extraneous cruft from Base64
    - enable dimming
 V0.1.2 - 2016/02/05
    - Added SSL Protocol support
    - authorization header rather than within URL
 V0.0.7 - 2016/02/05
    - Fixed Today counter and WindChill characteristic
 V0.0.6 - 2016/02/03
    - Full DarkSkies Virtual Sensor support (Rain, Wind, Barometer, Solar Radiation, Visibility
 V0.0.5 - 2016/02/03
    - Added YouLess counter support (Type: YouLess Meter, SubType: YouLess counter)
    - Expanded Temp sensor to include humidity & pressure (if present)
    - fixed UUID generation
 V0.0.4 - 2016/01/31
      - Fixed 'Siri Name' disappearance
 V0.0.3 - 2016/01/31
      - Added General Usage Sensors (Type: General, SubType: Percentage)
 V0.0.2 - 2016/01/31
      - Added Electric Consumption sensors (Type: General, SubType: kWh)
 V0.0.1 - 2016/01/31
      - Initial version
      - I make no claims to the quality of this shim. Function over form!


lib/mqtt.js
var mqtt = require('mqtt');
var platform;
var accessories = [];
var client;
var config = {host: "", port: 0, credentials: false, channel: ""};

module.exports = {
  Mqtt: Mqtt
}

function Mqtt(aPlatform, host, port, channel, credentials) {
  platform = aPlatform;
  platform.accessories(function(rawAccessories) {
    for (var i = 0; i < rawAccessories.length; i++)
    {
      var accessory = rawAccessories[i];
      accessories["idx_" + accessory.idx] = accessory;
    }

    config = {host: host, port: port, credentials: credentials, channel: channel};
    if (typeof config.credentials === 'undefined' || typeof config.credentials.username === 'undefined' || config.credentials.username.length == 0) {
      config.credentials = false;
    }
    this.connect();
  }.bind(this));
}

Mqtt.prototype.connect = function() {
    var connectOptions = {
    host: config.host,
    port: config.port
  };

  if (config.credentials)
  {
    connectOptions.username = config.credentials.username;
    connectOptions.password = config.credentials.password;
  }

  client = mqtt.connect(connectOptions);

  client.on('connect', function() {
    platform.forceLog("Successfully connected to MQTT broker.");
    client.subscribe(config.channel);
  });

  client.on('close', function(error) {
    platform.forceLog("Retrying connection to MQTT broker...");
  });

  client.on('error', function(error) {
    platform.forceLog("Could not connect to MQTT broker! (" + config.host + ":" + config.port + ")\n" + error);
  });

  client.on('message', function (topic, buffer) {
    var message = JSON.parse(buffer.toString());
    if (typeof message.nvalue !== 'undefined' || typeof message.svalue1 !== 'undefined') {
      var accessory = accessories["idx_" + message.idx];
      if (!accessory) {
        return;
      }

      accessory.handleMQTTMessage(message, function(characteristic, value) {
        if (typeof value !== 'undefined' && typeof characteristic !== 'undefined') {
          characteristic.setValue(value, null, "eDomoticz-MQTT");
        }
      });

    } else {
		  platform.log('[ERR] MQTT message received, but no nvalue or svalue1 was found:');
		  platform.log(message);
	  }
  });
}

Mqtt.prototype.send = function(message) {
  if (client)
  {
    var payload = message;
    if (typeof payload !== 'string') {
      payload = JSON.stringify(payload);
    }
    client.publish('domoticz/in', payload);
  }
}

lib/helper.js
var inherits = require('util').inherits;

module.exports = {
    Helper: Helper
}

function Helper() {

}

if (!Date.prototype.toISOString) {
  (function() {

    function pad(number) {
      if (number < 10) {
        return '0' + number;
      }
      return number;
    }

    Date.prototype.toISOString = function() {
      return this.getUTCFullYear() +
        '-' + pad(this.getUTCMonth() + 1) +
        '-' + pad(this.getUTCDate()) +
        'T' + pad(this.getUTCHours()) +
        ':' + pad(this.getUTCMinutes()) +
        ':' + pad(this.getUTCSeconds()) +
        '.' + (this.getUTCMilliseconds() / 1000).toFixed(3).slice(2, 5) +
        'Z';
    };

  }());
}

Date.prototype.addMinutes = function(h) {
  this.setTime(this.getTime() + (h*60*1000));
  return this;
};

Helper.sortByKey = function(array, key) {
    return array.sort(function(a, b) {
        var x = a[key];
        var y = b[key];
        return ((x < y) ? -1 : ((x > y) ? 1 : 0));
    });
};

Helper.oneDP = function(value) {
   var converted = value ? Math.round(value*10)/10 : 0;
   var fixed = converted.toFixed(1);
   return parseFloat(fixed);
};

Helper.cleanFloat = function(value) {
   var stringval = value ? value.toString() : "";
   stringval = stringval.replace(/[^0-9\.-]+/g,'')
   return parseFloat(stringval);
};

Helper.fixInheritance = function(subclass, superclass) {
    var proto = subclass.prototype;
    inherits(subclass, superclass);
    subclass.prototype.parent = superclass.prototype;
    for (var mn in proto) {
        subclass.prototype[mn] = proto[mn];
    }
};

Helper.HSVtoRGB = function(hsb) {
  var br = Math.round(hsb[2] / 100 * 254);
  var rgb = false;
  if (hsb[1] == 0){
    rgb = [br, br, br];
  } else {
    var hue = hsb[0] % 360;
    var f = hue % 60;
    var p = Math.round((hsb[2] * (100 - hsb[1])) / 10000 * 254);
    var q = Math.round((hsb[2] * (6000 - hsb[1] * f)) / 600000 * 254);
    var t = Math.round((hsb[2] * (6000 - hsb[1] * (60 - f))) / 600000 * 254);
    switch (Math.floor(hue / 60)){
      case 0: rgb = [br, t, p]; break;
      case 1: rgb = [q, br, p]; break;
      case 2: rgb = [p, br, t]; break;
      case 3: rgb = [p, q, br]; break;
      case 4: rgb = [t, p, br]; break;
      case 5: rgb = [br, p, q]; break;
    }
  }
  
  if (rgb)
  {
    var hex = "";
    for (var i = 0; i < 3; i++){
      var bit = (rgb[i] - 0).toString(16);
      hex += (bit.length == 1) ? '0' + bit : bit;
    }
    return hex;
  }

  return "FFFFFF";
};

lib/domoticz.js
var request = require("request");
var Helper = require('./helper.js').Helper;
var extend = require('util')._extend;

module.exports = {
    Domoticz: Domoticz
}

var baseHttpRequest = false;

function Domoticz() {

}

Domoticz.initialize = function(useSSL, requestHeaders) {
  var defaultRequestOptions = {
    headers: requestHeaders,
    json: true
  };

  if (useSSL)
  {
    defaultRequestOptions.agentOptions = {
      rejectUnauthorized: false
    };
  }

  baseHttpRequest = request.defaults(defaultRequestOptions);
};

Domoticz.devices = function(baseURL, roomID, completion, error) {
  if (baseHttpRequest === false) {
    return;
  }

  var url = baseURL + "type=devices&used=true&order=Name";
  if (roomID) {
      url += "&plan=" + roomID;
  }

  baseHttpRequest.get({
    url: url
  }, function(err, response, json) {
    if (!err && response.statusCode == 200 && json.result !== undefined)
    {
      var devices = [];
      var sArray = Helper.sortByKey(json.result, "Name");
      sArray.map(function(s) {
        devices.push(s);
      });

      if (typeof completion !== 'undefined' && completion !== false) {
        completion(devices);
      }
    }
    else
    {
      if (typeof error !== 'undefined' && error !== false) {
        error();
      }
    }
  });
};

Domoticz.deviceStatus = function(accessory, completion, error) {
  if (baseHttpRequest === false) {
    return;
  }

  var url = accessory.baseURL + "type=devices&rid=" + accessory.idx;
  baseHttpRequest.get({
    url: url
  }, function(err, response, json) {
    if (!err && response.statusCode == 200 && json.result !== undefined)
    {
      this.platform.log("Data Received for " + this.name + ".");
      if (typeof completion !== 'undefined' && completion !== false) {
        completion(json);
      }
    }
    else
    {
      this.platform.log("There was a problem connecting to Domoticz. (HTTP Status " + response.statusCode + ")\n- " + err);
      if (typeof error !== 'undefined' && error !== false) {
        error();
      }
    }
  }.bind(accessory));
};

Domoticz.updateDeviceStatus = function(accessory, command, parameters, completion) {
  if (accessory.platform.mqtt && this.isMQTTSupportedCommand(command, parameters))
  {
    var message = {"command": command, "idx": parseInt(accessory.idx)};
    extend(message, parameters);

    accessory.platform.mqtt.send(message);
    if (typeof completion !== 'undefined' && completion !== false) {
      completion(true);
    }
    return;
  }

  var url = accessory.baseURL + "type=command&param=" + encodeURI(command) + "&idx=" + accessory.idx;

  for (var key in parameters) {
    url += "&" + encodeURI(key) + "=" + encodeURI(parameters[key]);
  }

  Domoticz.updateWithURL(accessory, url, completion);
};

Domoticz.updateWithURL = function(accessory, url, completion) {
  if (baseHttpRequest === false) {
    return;
  }

  baseHttpRequest.put({
    url: url,
    header: this.requestHeaders,
    json: true
  },
  function(err, response) {
    var success = (typeof err === 'undefined' || !err);

    if (success) {
      this.platform.log(this.name + " sent command succesfully.");
    }
    else
    {
      this.platform.log("There was a problem sending command to " + this.name + ". (HTTP Status " + response.statusCode + ")\n- " + err);
      this.platform.log(response.body);
    }

    if (typeof completion !== 'undefined' && completion !== false) {
      completion(success);
    }
  }.bind(accessory));
};

Domoticz.isMQTTSupportedCommand = function(command, parameters) {
  if (command == "setcolbrightnessvalue") {
      return false;
  }

  return true;
}

lib/constants.js
global.Service;
global.Characteristic;
global.Types;
global.UUID;

module.exports = {
    DeviceTypeSwitch: 0,
    DeviceTypeDoorbell: 1,
    DeviceTypeContact: 2,
    DeviceTypeBlinds: 3,
    DeviceTypeSmoke: 5,
    DeviceTypeBlindsInverted: 6,
    DeviceTypeDimmer: 7,
    DeviceTypeMotion: 8,
    DeviceTypePushOn: 9,
    DeviceTypeDoorLock: 11,
    DeviceTypeBlindsPercentage: 13,
    DeviceTypeBlindsPercentageInverted: 16,
    DeviceTypeMedia: 17, //media - only supported as on/off switch at the moment.

    /* Made up, interal types */
    DeviceTypeSecuritySystem: 254, // Has HardwareTypeVal 67 w/o any SwitchTypeVal

    /*
    DeviceTypeNotSupportedYet: 4, //x10siren
    DeviceTypeNotSupportedYet: 9, //pushon
    DeviceTypeNotSupportedYet: 10, //pushoff
    DeviceTypeNotSupportedYet: 11, //doorlock
    DeviceTypeNotSupportedYet: 12, //dusk
    DeviceTypeNotSupportedYet: 14, //venetianus
    DeviceTypeNotSupportedYet: 15, //venetianeu
    */
}


index.js
//         ____                        _   _
//     ___|  _ \  ___  _ __ ___   ___ | |_(_) ___ ____
//    / _ | | | |/ _ \| '_ ` _ \ / _ \| __| |/ __|_  /
//   |  __| |_| | (_) | | | | | | (_) | |_| | (__ / /
//    \___|____/ \___/|_| |_| |_|\___/ \__|_|\___/___|
//       www.npmjs.com/package/homebridge-edomoticz
//
//   A Platform Plugin for HomeBridge by Marci & TheRamon
//           [http://twitter.com/marcisshadow]
//           [http://domoticz.com/forum/memberlist.php?mode=viewprofile&u=10884]
//
//     ** Remember to add platform to config.json **
//
// Example ~/.homebridge/config.json content:
//
// {
//  "bridge": {
//         "name": "Homebridge",
//         "username": "CC:21:3E:E4:DE:33", // << Randomize this...
//         "port": 51826,
//         "pin": "031-45-154",
//      },
//
//  "platforms": [{
//         "platform": "eDomoticz",
//         "name": "eDomoticz",
//         "server": "127.0.0.1",   // or "user:pass@ip"
//         "port": "8080",
//         "roomid": 0 ,  // 0 = all sensors, otherwise, room idx as shown at http://server:port/#/Roomplan
//         "ssl": 0,
//         "mqttenable": 1,
//         "mqttserver": "127.0.0.1",
//         "mqttport": "1883",
//         "mqttauth": 1,          //only needed if you've password protected mosquitto via mosquitto.conf
//         "mqttuser": "username", //only needed if you've password protected mosquitto via mosquitto.conf
//         "mqttpass": "password"  //only needed if you've password protected mosquitto via mosquitto.conf
//      }],
//
//  "accessories":[]
// }
//

var Domoticz = require('./lib/domoticz.js').Domoticz;
var Mqtt = require('./lib/mqtt.js').Mqtt;
var eDomoticzAccessory = require('./lib/domoticz_accessory.js');
var Constants = require('./lib/constants.js');
var Helper = require('./lib/helper.js').Helper;
var eDomoticzServices = require('./lib/services.js').eDomoticzServices;
const util = require('util');

module.exports = function(homebridge) {
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  Types = homebridge.hapLegacyTypes;
  UUID = homebridge.hap.uuid;

  Helper.fixInheritance(eDomoticzServices.TotalConsumption, Characteristic);
  Helper.fixInheritance(eDomoticzServices.CurrentConsumption, Characteristic);
  Helper.fixInheritance(eDomoticzServices.GasConsumption, Characteristic);
  Helper.fixInheritance(eDomoticzServices.TempOverride, Characteristic);
  Helper.fixInheritance(eDomoticzServices.MeterDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.GasDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.CurrentUsage, Characteristic);
  Helper.fixInheritance(eDomoticzServices.UsageDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.TodayConsumption, Characteristic);
  Helper.fixInheritance(eDomoticzServices.Barometer, Characteristic);
  Helper.fixInheritance(eDomoticzServices.WaterFlow, Characteristic);
  Helper.fixInheritance(eDomoticzServices.TotalWaterFlow, Characteristic);
  Helper.fixInheritance(eDomoticzServices.WaterDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.WeatherService, Service);
  Helper.fixInheritance(eDomoticzServices.WindSpeed, Characteristic);
  Helper.fixInheritance(eDomoticzServices.WindChill, Characteristic);
  Helper.fixInheritance(eDomoticzServices.WindDirection, Characteristic);
  Helper.fixInheritance(eDomoticzServices.WindDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.Rainfall, Characteristic);
  Helper.fixInheritance(eDomoticzServices.RainDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.Visibility, Characteristic);
  Helper.fixInheritance(eDomoticzServices.VisibilityDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.SolRad, Characteristic);
  Helper.fixInheritance(eDomoticzServices.SolRadDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.LocationService, Service);
  Helper.fixInheritance(eDomoticzServices.Location, Characteristic);
  Helper.fixInheritance(eDomoticzServices.InfotextDeviceService, Service);
  Helper.fixInheritance(eDomoticzServices.Infotext, Characteristic);
  homebridge.registerAccessory("homebridge-edomoticz", "eDomoticz", eDomoticzAccessory);
  homebridge.registerPlatform("homebridge-edomoticz", "eDomoticz", eDomoticzPlatform);
};

function eDomoticzPlatform(log, config, api) {
  this._cachedAccessories = false;
  this.forceLog = log;
  this.log = function() {
      if (typeof process.env.DEBUG !== 'undefined') {
          log(util.format.apply(this, arguments));
      }
  };

  this.config = config;
  this.server = config.server;
  this.authorizationToken = false;
  if (this.server.indexOf(":") > -1 && this.server.indexOf("@") > -1)
  {
    tmparr = this.server.split("@");
    var Base64 = {
        _keyStr: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",
        encode: function(e) {
            var t = "";
            var n, r, i, s, o, u, a;
            var f = 0;
            e = Base64._utf8_encode(e);
            while (f < e.length) {
                n = e.charCodeAt(f++);
                r = e.charCodeAt(f++);
                i = e.charCodeAt(f++);
                s = n >> 2;
                o = (n & 3) << 4 | r >> 4;
                u = (r & 15) << 2 | i >> 6;
                a = i & 63;
                if (isNaN(r)) {
                    u = a = 64;
                } else if (isNaN(i)) {
                    a = 64;
                }
                t = t + this._keyStr.charAt(s) + this._keyStr.charAt(o) + this._keyStr.charAt(u) + this._keyStr.charAt(a);
            }
            return t;
        },
        _utf8_encode: function(e) {
            e = e.replace(/\r\n/g, "\n");
            var t = "";
            for (var n = 0; n < e.length; n++) {
                var r = e.charCodeAt(n);
                if (r < 128) {
                    t += String.fromCharCode(r);
                } else if (r > 127 && r < 2048) {
                    t += String.fromCharCode(r >> 6 | 192);
                    t += String.fromCharCode(r & 63 | 128);
                } else {
                    t += String.fromCharCode(r >> 12 | 224);
                    t += String.fromCharCode(r >> 6 & 63 | 128);
                    t += String.fromCharCode(r & 63 | 128);
                }
            }
            return t;
        }
    };
    this.authorizationToken = Base64.encode(tmparr[0]);
    this.server = tmparr[1];
  }
  this.mqttenable = config.mqttenable;
  this.ssl = (config.ssl == 1);
  this.port = config.port;
  this.room = config.roomid;
  this.api = api;
  this.mqtt = false;

  var requestHeaders = {};
  if (this.authorizationToken) {
    requestHeaders['Authorization'] = 'Basic ' + this.authorizationToken;
  }
  Domoticz.initialize(this.ssl, requestHeaders);

  if (config.mqttenable===1 && this.api)
  {
    this.api.once("domoticzAccessoriesLoaded", function() {
      this.accessories(function(accessories) {
        if (accessories.length > 0) {
          this.mqtt = new Mqtt(this, config.mqttserver, config.mqttport, 'domoticz/out', {username: config.mqttuser, password: config.mqttpass});
        }
      }.bind(this));
    }.bind(this));
  }
}

eDomoticzPlatform.prototype = {
  accessories: function(callback, forceUpdate) {
    if (this._cachedAccessories && ((typeof forceUpdate === 'undefined') || !forceUpdate)) {
      callback(this._cachedAccessories);
      return;
    }

    this.log("Fetching Domoticz lights and switches...");

    var baseUrl = "http" + (this.ssl ? "s" : "") + "://" + this.server + ":" + this.port + "/json.htm?";
    Domoticz.devices(baseUrl, this.room, function(devices) {
      this.log("You have " + devices.length + " devices defined in Domoticz.");

      var newAccessories = [];
      for (var i = 0; i < devices.length; i++)
      {
        var device = devices[i];

        var accessory = new eDomoticzAccessory(this, baseUrl, false, device.Used, device.idx, device.Name, device.HaveDimmer, device.MaxDimLevel, device.SubType, device.Type, device.BatteryLevel, device.SwitchType, device.SwitchTypeVal, device.HardwareTypeVal, this.eve);
        newAccessories.push(accessory);
      }

      this._cachedAccessories = newAccessories;
      callback(this._cachedAccessories);
      this.api.emit("domoticzAccessoriesLoaded");

      if (this._cachedAccessories.length == 0)
      {
        if (this.room == 0) {
          this.forceLog("You do not have any Domoticz devices yet. Please add some devices and restart HomeBridge.");
        }
        else {
          this.forceLog("You do not have any Domoticz devices in this room (roomid: " + this.room + ") yet. Please add some devices to this room and restart HomeBridge.");
        }
      }
    }.bind(this), function() {
      this.forceLog("There was a problem connecting to Domoticz.");
    }.bind(this));
  }
};
